src/logging.rs

```diff
@@ -0,0 +1,140 @@
+use std::fs::File;
+use std::io::Write;
+use std::os::unix::io::FromRawFd;
+use std::ffi::CString;
+use std::os::unix::net::UnixDatagram;
+use std::os::unix::io::{AsRawFd, IntoRawFd};
+use nix::sys::socket::{ControlMessage, MsgFlags, sendmsg};
+use nix::sys::memfd::MemFdCreateFlag;
+use nix::sys::memfd::memfd_create;
+use nix::fcntl::*;
+
+use crate::errors::*;
+
+/// Default path of the journald AF_UNIX datagram socket.
+const SD_JOURNAL_SOCK_PATH: &str = "/run/systemd/journal/socket";
+
+/// Log priority values.
+///
+/// See `man 3 syslog`.
+#[repr(u8)]
+pub enum Priority {
+    /// system is unusable
+    Emergency = 0,
+    /// action must be taken immediately
+    Alert,
+    /// critical conditions
+    Critical,
+    /// error conditions
+    Error,
+    /// warning conditions
+    Warning,
+    /// normal, but significant, condition
+    Notice,
+    /// informational message
+    Info,
+    /// debug-level message
+    Debug,
+}
+
+#[inline(always)]
+fn is_valid_char(c: char) -> bool {
+    c.is_uppercase() || c.is_numeric() || c == '_'
+}
+
+/// The variable name must be in uppercase and consist only of characters, 
+/// numbers and underscores, and may not begin with an underscore.
+fn is_valid_field(input: &str) -> bool {
+    if input.is_empty() {
+        return false;
+    }
+
+    if !input.chars().all(is_valid_char) {
+        return false;
+    }
+
+    if input.starts_with('_') {
+        return false;
+    }
+
+    true
+}
+
+fn add_field_and_payload(data: &mut String, field: &str, payload: &str) {
+    if is_valid_field(field) {
+        let field_payload = format!("{}={}\n", field, payload);
+        data.push_str(&field_payload);
+    }
+}
+
+/// Print a message to the systemd journal with the given priority.
+pub fn sd_journal_print(priority: Priority, msg: &str) -> Result<()> {
+    let sock = UnixDatagram::unbound()?;
+    sock.connect(SD_JOURNAL_SOCK_PATH)?;
+
+    let mut data = String::new();
+    add_field_and_payload(&mut data, "PRIORITY", &(priority as u8).to_string());
```

Comment:
`u8::from(priority)` should work here too, and it is a bit better as it avoids a blind cast.

Reply:
I will have to implement `std::convert::From` for `Priority` if you want to use `u8::from()` here

Reply:
Interesting, I thought we got that for free via `repr(u8)`.
Anyway, I think it makes to have the `Priority::Into<u8>` implemented.